### PR TITLE
78 producer mode

### LINKED ISSUE #78: Add a producer for json flatbuffer messages
Add an option to run in producer mode and pass a json file to be published using the `json` flatbuffer schema: https://github.com/ess-dmsc/streaming-data-types/blob/master/schemas/json_json.fbs

### PR DESCRIPTION
**Pull request by mchorazak**
### Description of work

Produce messages from file to kafka using json schema. 

run kafkacow with `-P -b broker -t TopicToProduceTo -f /home/user/Example.json` to send json file using json schema.

Producer and Consumer use factory method.
`RequestHandler` is responsible for all three modes.
`FlatbufferTranslator` contains both serializing(producing) and deserializing(consuming) methods.
All classes responsible for communication with Kafka were moved to KafkaW namespace, some of them were excluded from codecov.

CMake code duplication reduced by creating a list of dependencies in the top-level CMake.

### Issue

Closes #78 

### Unit Tests

One test that covers producer's path in `RequestHandler`.
Two tests covering argument checking.

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/RequestHandler.cpp [hunk 1]
```diff
@@ -28,24 +27,29 @@ void RequestHandler::checkAndRun() {
...
+    checkIfTopicEmpty(UserArguments.TopicName);
     if (UserArguments.GoBack > -2 && UserArguments.OffsetToStart > -2) {
-      subscribeAndConsume(UserArguments.Name, UserArguments.GoBack,
+      subscribeAndConsume(UserArguments.TopicName, UserArguments.GoBack,
                           UserArguments.PartitionToConsume,
                           UserArguments.OffsetToStart);
     } else {
       if (UserArguments.OffsetToStart > -2) {
-        subscribeAndConsume(UserArguments.Name, UserArguments.OffsetToStart);
+        subscribeAndConsume(UserArguments.TopicName,
```
Comment (matthew-d-jones) [inline src/RequestHandler.cpp]: codecov says this line is not covered, it this a scenario which should have a unit test?

File: src/KafkaW/Message.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace KafkaW {
+
+class Message {
+public:
+  explicit Message(flatbuffers::DetachedBuffer InputBuffer)
+      : Buffer(std::move(InputBuffer)) {}
+
+  char *data() { return reinterpret_cast<char *>(Buffer.data()); }
+  size_t size() { return Buffer.size(); }
```
Comment (matthew-d-jones) [inline src/KafkaW/Message.h]: The methods of `Message` are not covered by unit tests. Perhaps worth adding a test where you serialise a simple message, put it in a Message, and then check you can get it back out and deserialise it?

File: src/FlatbuffersTranslator.cpp [hunk 3]
```diff
@@ -18,6 +20,10 @@ FlatbuffersTranslator::deserializeToJSON(KafkaMessageMetadataStruct MessageData,
     FileID.clear();
   }
 
+  if (FileID == "json") {
+    return GetJsonData(MessageData.Payload.c_str())->json()->str();
```
Comment (matthew-d-jones) [inline src/FlatbuffersTranslator.cpp]: This line isn't covered by unit tests. Can/should it be?

File: test/RequestHandlerTest.cpp [hunk 4]
```diff
@@ -16,42 +17,39 @@ class RequestHandlerTest : public ::testing::Test {};
 
 TEST(RequestHandlerTest,
      subscribe_consume_n_last_messages_throws_if_incorrect_arguments_test) {
-  auto KafkaConnection = std::make_unique<ConnectKafkaFake>(ConnectKafkaFake());
   UserArgumentStruct UserArguments;
-  RequestHandler NewRequestHandler(std::move(KafkaConnection), UserArguments,
-                                   updateSchemas(false));
+  UserArguments.ConsumerMode = true;
+  RequestHandler NewRequestHandler(UserArguments, updateSchemas(false), false);
```
Comment (matthew-d-jones) [inline test/RequestHandlerTest.cpp]: Putting this `false` in an appropriately named `const bool` in the anonymous namespace would help readability of the tests. I don't think the `SchemaPath` variable is used any more so you can remove that whilst you are there :)

File: src/RequestHandler.cpp (hunk 1 above)
Comment (mchorazak) [reply src/RequestHandler.cpp]: The way this scenario is tested changed after we introduced that infinite loop. Testing starts with method that is called by this `subscribeAndConsume` which has a boolean preventing the innfinite loop from running. Starting testing here makes tests hang.

**Pull request merged**